### PR TITLE
Refactor health checks

### LINKED ISSUE #1941: Refactor health checks
#### Is your feature request related to a problem?

Since the health check actually relies on the state of underlying storage, it's not suitable for both liveness and readiness check in Kubernetes.

#### Describe the solution you'd like to see

Add separate liveness and readiness health checks.

Move health checks to the telemetry server, because it's only needed internally.

Liveness check should always just return "ok", readiness check should be what the health check is at the moment.

Consider using [go-sundheit](https://github.com/AppsFlyer/go-sundheit) (beware that it imports a couple heavy/unnecessary dependencies). It's also in line with #1386 and #1397

#### Additional context

#853 points out that the current health check causes Dex to fail when the underlying storage is unavailable. Separating liveness and readiness checks would definitely be an improvement, but might not be enough to improve that situation. Although readiness checks should clearly signal when a pod is unable to serve traffic (and in case of a failing/unavailable storage that is the case), Dex is not HA at the moment (there are some GC jobs internally that should not run in parallel), so taking the single pod out of traffic would make Dex become completely unavailable.

### PR DESCRIPTION
**Pull request by sagikazarmark**
#### Overview

Improve health checks by separating liveness and readiness checks

#### What this PR does / why we need it

Closes #1941

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: storage/health.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package storage
+
+import (
+	"fmt"
+	"time"
+)
+
+// NewCustomHealthCheckFunc returns a new health check function.
+func NewCustomHealthCheckFunc(s Storage, now func() time.Time) func() (details interface{}, err error) {
```
Comment (bonifaido) [inline storage/health.go:9]: So if I'm correct, the health check creates and deletes an auth request every time it hits the check endpoint?

Comment (sagikazarmark) [reply storage/health.go:9]: It does. It's not optimal IMO, but this PR is about moving the health check to the telemetry endpoint.

Most storage solutions implement some sort of Ping under the hood, so we should probably use that in the future.

Comment (sagikazarmark) [reply storage/health.go:9]: Sorry, not quite. The health check runs in the background with a predefined interval, the health endpoint just reports back the last result. There is no behavior change on this front.

But creating dummy auth requests is probably not a good idea. I'll try to address that in a followup PR, but I need this change for the helm chart.

Comment (bonifaido) [reply storage/health.go:9]: No problem, I'm okay with this right now.

**Review by bonifaido — APPROVED** [review]

**Pull request merged**